oneflow/user/kernels/hardtanh_kernel.cpp

```diff
@@ -0,0 +1,108 @@
+
+namespace oneflow {
+
+namespace user_op {
+
+template<DeviceType device_type, typename T>
+class CpuHardtanhKernel final : public OpKernel {
+ public:
+  CpuHardtanhKernel() = default;
+  ~CpuHardtanhKernel() = default;
+
+ private:
+  void Compute(KernelComputeContext* ctx) const override {
+    const Tensor* in_tensor = ctx->Tensor4ArgNameAndIndex("in", 0);
+    Tensor* out_tensor = ctx->Tensor4ArgNameAndIndex("out", 0);
+    const T min_val = static_cast<T>(ctx->Attr<double>("min_val"));
+    const T max_val = static_cast<T>(ctx->Attr<double>("max_val"));
+    const T* in_ptr = in_tensor->dptr<T>();
+    T* out_ptr = out_tensor->mut_dptr<T>();
+
+    const int32_t elem_cnt = in_tensor->shape().elem_cnt();
+    FOR_RANGE(int32_t, i, 0, elem_cnt) {
+      if (in_ptr[i] > max_val)
+        out_ptr[i] = max_val;
+      else if (in_ptr[i] < min_val)
+        out_ptr[i] = min_val;
+      else
+        out_ptr[i] = in_ptr[i];
+    }
+  }
+  bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
+};
+
+#define REGISTER_CPU_HARDTANH_KERNEL(device, dtype)                                       \
+  REGISTER_USER_KERNEL("hardtanh")                                                        \
+      .SetCreateFn<CpuHardtanhKernel<device, dtype>>()                                    \
+      .SetIsMatchedHob((HobDeviceTag() == device)                                         \
+                       & (HobDataType("out", 0) == GetDataType<dtype>::value))            \
+      .SetInplaceProposalFn(                                                              \
+          [](const InferContext&, AddInplaceArgPair AddInplaceArgPairFn) -> Maybe<void> { \
+            OF_RETURN_IF_ERROR(AddInplaceArgPairFn("out", 0, "in", 0, true));             \
+            return Maybe<void>::Ok();                                                     \
+          });
+
+REGISTER_CPU_HARDTANH_KERNEL(DeviceType::kCPU, float)
+REGISTER_CPU_HARDTANH_KERNEL(DeviceType::kCPU, double)
+
+template<DeviceType device_type, typename T>
+class CpuHardtanhGradKernel final : public OpKernel {
+ public:
+  CpuHardtanhGradKernel() = default;
+  ~CpuHardtanhGradKernel() = default;
+
+ private:
+  void Compute(KernelComputeContext* ctx) const override {
+    const Tensor* y_tensor = ctx->Tensor4ArgNameAndIndex("y", 0);
+    const Tensor* dy_tensor = ctx->Tensor4ArgNameAndIndex("dy", 0);
+    Tensor* dx_tensor = ctx->Tensor4ArgNameAndIndex("dx", 0);
+    const T* y_ptr = y_tensor->dptr<T>();
+    const T* dy_ptr = dy_tensor->dptr<T>();
+    T* dx_ptr = dx_tensor->mut_dptr<T>();
+
+    const T min_val = static_cast<T>(ctx->Attr<double>("min_val"));
+    const T max_val = static_cast<T>(ctx->Attr<double>("max_val"));
+    const T zero_t = static_cast<T>(0);
+
+    const int32_t elem_cnt = y_tensor->shape().elem_cnt();
+    FOR_RANGE(int32_t, i, 0, elem_cnt) {
+      dx_ptr[i] = (y_ptr[i] > min_val && y_ptr[i] < max_val) ? dy_ptr[i] : zero_t;
```

Comment:
这个求导公式我有疑问，hardtanh的公式是：
```
           min_val  (x < min_val)
y(x) =   x
           max_val (x > max_val)
```
那么对应的梯度应该是：
```
           0(x < min_val)
y(x) =   1
           0(x > max_val)
```

所以反向求梯度应该是：
```
           0*dy=0(x < min_val)
y(x) =   1*dy = dy
           0*dy = 0(x > max_val)
```

所以，反向算子，应该拿到的是 `x`(而不是 `y`) 和 `dy` 吧？

Reply:
我这里是做了inplace操作, 因为Pytorch的hardtanh支持inplace. 说人话就是原地操作, hardtanh的输出直接在输入上原地修改. **因此Inplace操作要求后向传播 不依赖于输入值,**, hardtanh在边界处没有梯度, 所以是可以通过最后的输出y, 来判断他是否在 [min_val, max_val] 这个范围区间的.  做Inplace操作的好处是节约内存.

Reply:
因为inplace这样写是有道理的，不过既然这样，是不是
```suggestion
      dx_ptr[i] = (y_ptr[i] != min_val && y_ptr[i] != max_val) ? dy_ptr[i] : zero_t;
```
这样更可以表达，你因为要inplace，而用y来判断梯度公式